### PR TITLE
ISSUE-74 Modularize TypeState

### PR DESCRIPTION
**Pull request by quantranhong1999**
Working on rebasing...
Done

### CODE REVIEW

**Comment by chibenwa** [comment]
We need to solve the conflicts on this :-)

**Comment by quantranhong1999** [comment]
Need submodule to be updated on master.

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterStateChangeTest.scala [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+import sttp.client3.{Identity, RequestT, SttpBackend, asWebSocket, basicRequest}
+import sttp.model.Uri
+import sttp.monad.MonadError
+import sttp.monad.syntax.MonadErrorOps
+import sttp.ws.WebSocketFrame
+import sttp.ws.WebSocketFrame.Text
+
+import java.net.URI
+
+class FilterTypeNameModule extends AbstractModule {
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterStateChangeTest.scala:29]: Arf again Quan these bindings are not specific to tests, you need to move it to the top level maven module declaration!

Overwize we will not have these modules in production!

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterStateChangeTest.scala [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            ws.send(WebSocketFrame.text(
+              """{
+                |  "@type": "WebSocketPushEnable",
+                |  "dataTypes": ["FilterTypeName"]
+                |}""".stripMargin))
+
+            Thread.sleep(100)
+
+            server.getProbe(classOf[JmapEventBusProbe])
+              .emitStateChange(stateChangeEvent, accountId)
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterStateChangeTest.scala:77]: I would prefer to have a real `Filter/set` request.

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterStateChangeTest.scala [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                   |	"changed": {
+                   |		"29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6": {
+                   |			"FilterTypeName": "1"
+                   |		}
+                   |	}
+                   |}""".stripMargin)
+  }
+
+  @Test
+  def pushShouldFailWhenWrongFilterTypeName(server: GuiceJamesServer): Unit = {
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterStateChangeTest.scala:140]: We do not care about this test you can remove it.

File: tmail-backend/integration-tests/jmap/memory-jmap-integration-tests/src/test/java/com/linagora/tmail/james/MemoryLinagoraFilterStateChangeTest.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+public class MemoryLinagoraFilterStateChangeTest implements LinagoraFilterStateChangeTest {
+    @RegisterExtension
+    static JamesServerExtension jamesServerExtension = new JamesServerBuilder<MemoryConfiguration>(tmpDir ->
+        MemoryConfiguration.builder()
+            .workingDirectory(tmpDir)
+            .configurationFromClasspath()
+            .build())
+        .server(configuration -> MemoryServer.createServer(configuration)
+            .overrideWith(new TestJMAPServerModule())
+            .overrideWith(new FilterTypeNameModule()))
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/memory-jmap-integration-tests/src/test/java/com/linagora/tmail/james/MemoryLinagoraFilterStateChangeTest.java:22]: FilterTypeNameModule should be part of the top level module declaration, tests should not need to specify it.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/FilterSerializer.scala [hunk 5]
```diff
@@ -36,6 +35,13 @@ case class FilterSerializer @Inject()(mailboxIdFactory: MailboxId.Factory) {
   implicit val actionReads: Reads[Action] = Json.reads[Action]
   implicit val ruleWithIdReads: Reads[RuleWithId] = Json.reads[RuleWithId]
   implicit val updateReads: Reads[Update] = Json.valueReads[Update]
+  implicit val filterStateReads: Reads[FilterState] = {
+    case JsString(string) => FilterState.parse(string).fold(
+      error => JsError(error.getMessage),
+      filterState => JsSuccess(filterState)
+    )
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/FilterSerializer.scala:42]: I personally do not like lonely `)` brackets alone on their lines.... Can you move that one up?

**Review by chibenwa — COMMENTED** [review]
Appart from the guice module declaration, (which **should be part of the FilterMethodModule!!!**) it looks good to me.

**Comment by Arsnael** [comment]
> Need submodule to be updated on master.

Been updated and merged. Rebase please

**Review by Arsnael — COMMENTED** [review]
Nothing to add, you can get rid of the last commit after you rebase, as I had to do this anyway when merging with latest james project

File: tmail-backend/integration-tests/jmap/distributed-jmap-integration-tests/src/test/java/com/linagora/tmail/james/DistributedLinagoraFilterStateChangeTest.java [hunk 6]
```diff
@@ -35,6 +36,16 @@
         .extension(new AwsS3BlobStoreExtension())
         .server(configuration -> DistributedServer.createServer(configuration)
             .overrideWith(new TestJMAPServerModule())
-            .overrideWith(new FilterTypeNameModule()))
+            .overrideWith(new JmapGuiceFilterStateChangeModule()))
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/distributed-jmap-integration-tests/src/test/java/com/linagora/tmail/james/DistributedLinagoraFilterStateChangeTest.java:39]: We should not need any additional module here

Comment (quantranhong1999) [reply tmail-backend/integration-tests/jmap/distributed-jmap-integration-tests/src/test/java/com/linagora/tmail/james/DistributedLinagoraFilterStateChangeTest.java:39]: oops sorry i forgot

File: tmail-backend/integration-tests/jmap/memory-jmap-integration-tests/src/test/java/com/linagora/tmail/james/MemoryLinagoraFilterStateChangeTest.java [hunk 7]
```diff
@@ -19,6 +20,16 @@
             .build())
         .server(configuration -> MemoryServer.createServer(configuration)
             .overrideWith(new TestJMAPServerModule())
-            .overrideWith(new FilterTypeNameModule()))
+            .overrideWith(new JmapGuiceFilterStateChangeModule()))
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/memory-jmap-integration-tests/src/test/java/com/linagora/tmail/james/MemoryLinagoraFilterStateChangeTest.java:23]: Idem

File: tmail-backend/integration-tests/jmap/distributed-jmap-integration-tests/src/test/java/com/linagora/tmail/james/DistributedLinagoraFilterStateChangeTest.java (hunk 6 above)
Comment (chibenwa) [reply tmail-backend/integration-tests/jmap/distributed-jmap-integration-tests/src/test/java/com/linagora/tmail/james/DistributedLinagoraFilterStateChangeTest.java:39]: Please **do not put your bindings here**. We need them in production.

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/module/JmapGuiceFilterStateChangeModule.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+import com.google.inject.AbstractModule;
+import com.google.inject.multibindings.Multibinder;
+
+public class JmapGuiceFilterStateChangeModule extends AbstractModule {
+    @Override
+    protected void configure() {
+        Multibinder.newSetBinder(binder(), GuiceProbe.class)
+            .addBinding()
+            .to(JmapEventBusProbe.class);
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/module/JmapGuiceFilterStateChangeModule.java:33]: Do that in FilterSetMethodModule

File: tmail-backend/integration-tests/jmap/memory-jmap-integration-tests/src/test/java/com/linagora/tmail/james/MemoryLinagoraFilterStateChangeTest.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+public class MemoryLinagoraFilterStateChangeTest implements LinagoraFilterStateChangeTest {
+    @RegisterExtension
+    static JamesServerExtension jamesServerExtension = new JamesServerBuilder<MemoryConfiguration>(tmpDir ->
+        MemoryConfiguration.builder()
+            .workingDirectory(tmpDir)
+            .configurationFromClasspath()
+            .build())
+        .server(configuration -> MemoryServer.createServer(configuration)
+            .overrideWith(new TestJMAPServerModule())
+            .overrideWith(new JmapGuiceFilterStateChangeModule()))
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/memory-jmap-integration-tests/src/test/java/com/linagora/tmail/james/MemoryLinagoraFilterStateChangeTest.java:23]: Idem remove this declaration from here

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/module/JmapGuiceFilterStateChangeModule.java (hunk 8 above)
Comment (quantranhong1999) [reply tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/module/JmapGuiceFilterStateChangeModule.java]: I don't know why the commit that dropping this missed after rebase. My bad.

**Review by chibenwa — APPROVED** [review]

**Comment by Arsnael** [comment]
Squash your last fixup and we can merge it @quantranhong1999 :)

**Pull request merged**